langfuse/integrations.py

```diff
@@ -0,0 +1,99 @@
+
+
+load_dotenv()
+
+
+class CreateArgsExtractor:
+    def __init__(self, name=None, **kwargs):
+        self.args = {}
+        self.args["name"] = name
+        self.kwargs = kwargs
+
+    def get_langfuse_args(self):
+        return {**self.args, **self.kwargs}
+
+    def get_openai_args(self):
+        return self.kwargs
+
+
+class OpenAILangfuse:
+    def __init__(self):
+        self.langfuse = Langfuse(os.environ["LF_PK"], os.environ["LF_SK"], os.environ["HOST"])
+
+    def _get_call_details(self, result, **kwargs):
+        name = kwargs.get("name", "OpenAI-generation")
+        if result.object == "chat.completion":
+            prompt = kwargs.get("messages", [{}])[-1].get("content", "")
+            completion = result.choices[-1].message.content
+        elif result.object == "text_completion":
+            prompt = kwargs.get("prompt", "")
+            completion = result.choices[-1].text
+        else:
+            completion = ""
+        model = result.model
+        usage = None if result.usage is None else LlmUsage(**result.usage)
+        endTime = datetime.now()
+        modelParameters = {
+            "temperature": kwargs.get("temperature", 1),
+            "maxTokens": kwargs.get("max_tokens", float("inf")),
+            "top_p": kwargs.get("top_p", 1),
+            "frequency_penalty": kwargs.get("frequency_penalty", 0),
+            "presence_penalty": kwargs.get("presence_penalty", 0),
+        }
+        all_details = {
+            "name": name,
+            "prompt": prompt,
+            "completion": completion,
+            "endTime": endTime,
+            "model": model,
+            "modelParameters": modelParameters,
+            "usage": usage,
+        }
+        return all_details
+
+    def _log_result(self, result, call_details):
+        generation = InitialGeneration(**call_details)
+        self.langfuse.generation(generation)
+        self.langfuse.flush()
+        return result
+
+    def langfuse_modified(self, func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            try:
+                startTime = datetime.now()
+                arg_extractor = CreateArgsExtractor(*args, **kwargs)
+                result = func(**arg_extractor.get_openai_args())
+                call_details = self._get_call_details(result, **arg_extractor.get_langfuse_args())
+                call_details["startTime"] = startTime
+            except Exception as ex:
+                raise ex
+
+            return self._log_result(result, call_details)
+
+        return wrapper
+
+    def replace_openai_funcs(self):
+        api_resources_classes = [
+            (ChatCompletion, "create"),
+            (Completion, "create"),
+        ]
+
+        for api_resource_class, method in api_resources_classes:
+            create_method = getattr(api_resource_class, method)
+            setattr(api_resource_class, method, self.langfuse_modified(create_method))
+
+
+modifier = OpenAILangfuse()
```

Comment:
Do you think we can instantiate this as a Singleton? I fear our users will import this multiple times and then create multiple Langfuse objects under the hood. 
I think one instantiation of the Langfuse object would be ideal. Could we also expose the flush function from there? If yes, we can remove it above in the `_log_result` method and thereby save latencies for our users

Reply:
Other than this, i think we are good to go. I just tested it on my machine also - seems to work well :)

Reply:
<img width="501" alt="image" src="https://github.com/langfuse/langfuse-python/assets/17686849/250d1f12-3c42-4995-bc76-c65b8afab7ba">

`create` is of type any here. If you import `import openai` into `integrations.py`, this should be typed. Could you add that?

Reply:
Thanks a lot for the change on the singleton! What do you think of concurrency? I just read this here: https://stackabuse.com/creating-a-singleton-in-python/. 
Also, can you add a flush functionality to the module? Otherwise we hide the langfuse object and our users are not able to flush before the program exits.

Reply:
This could help here https://medium.com/analytics-vidhya/how-to-create-a-thread-safe-singleton-class-in-python-822e1170a7f6. I have made the change

Reply:
I don't find a way to directly calling `flush` using openai. In order to use it we need to create a object of `OpenAILangfuse`. Do you need to do that??

Reply:
```
    langfuse = Langfuse()
    completion = openai.ChatCompletion.create(
        langfuse=langfuse, model="gpt-3.5-turbo", messages=[{"role": "user", "content": "1 + 1 = "}], temperature=0
    )
    langfuse.flush()
```

What do you think about such an experience? If we do that, we can also allow more complex scenarios in the future:
```
    langfuse = Langfuse()

    span = langfuse.span(...)

    completion = openai.ChatCompletion.create(
        langfuse=span, model="gpt-3.5-turbo", messages=[{"role": "user", "content": "1 + 1 = "}], temperature=0
    )
    langfuse.flush()
```

What do you think about that?

Reply:
```
    langfuse = Langfuse()
    completion = openai.ChatCompletion.create(
        langfuse=langfuse, model="gpt-3.5-turbo", messages=[{"role": "user", "content": "1 + 1 = "}], temperature=0
    )
    langfuse.flush()
```


What do you think about such an experience? If we do that, we can also allow more complex scenarios in the future:
```
    langfuse = Langfuse()

    span = langfuse.span(...)

    completion = openai.ChatCompletion.create(
        langfuse=span, model="gpt-3.5-turbo", messages=[{"role": "user", "content": "1 + 1 = "}], temperature=0
    )
    langfuse.flush()
```
What do you think about that?

Reply:
@Dev-Khant sorry if the above was not shown to you. What do you think of the above?

Reply:
Sorry @maxdeichmann didn't saw the above comment. It's a useful approach but the `languse` inside openai wrapper will flush the inner processes and while the outer `langfuse` will handle processes defined out the class. For eg. processes like as you showed here `span = langfuse.span(...)`. So there won't be any conflict with `flush()` being used by two different instances of languse, right?

Reply:
No worries :)

Yes, you can create two different `Langfuse` objects in theory, they wont interact with each other. However, each of these will create their own worker thread to take care of network requests. I dont think this would be great for the performance of our users applications.
Also, the above approach gives us way more flexibility using the integration. Users could create generations in sub-spans of their trace which would not be possible without.

Reply:
@maxdeichmann Sure then we could go with this approach!

Reply:
I think before merging we should have that in here. I can take that over on the weekend.

Reply:
Sure, let me know if you need my help.